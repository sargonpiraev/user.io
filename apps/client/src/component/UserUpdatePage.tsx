import * as React from 'react'
import { connect } from 'react-redux'
import { Button, Card, Container, Divider, Header, Icon, Image, Input, Loader } from 'semantic-ui-react'
import ProgressiveImage from 'react-progressive-image'

import { userNameChange } from './../store/action/userNameChange'
import { userNameEditorEnable } from './../store/action/userNameEditorEnable'
import { userReview } from './../store/action/userReview'
import { userUpdate } from './../store/action/userUpdate'
import { WithRouterProps } from 'react-router'
import { IState } from '../index'

import { placeholder } from './User'

export class UserUpdatePage extends React.Component<any, any> {

	constructor(props: any) {
		super(props)
		this.onNameChange = this.onNameChange.bind(this)
		this.onEditNameClick = this.onEditNameClick.bind(this)
		this.onNameUpdateClick = this.onNameUpdateClick.bind(this)
	}

	public render() {
		return (
			<Container style={ { padding: '1rem 0' } }>
				<Header as='h1'>User update page</Header>
				<Divider />
				{ this.props.user ? this.renderUser() : this.renderLoader() }
			</Container>
		)
	}

	public componentDidMount() {
		this.props.userReview(parseInt(this.props.id, 10))
	}

	public renderName() {
		return (
			<div>
				<span id='name'>{ this.props.user.name }</span>
				&nbsp;
				<Icon name='pencil' onClick={ this.onEditNameClick } />
			</div>
		)
	}

	public renderNameEditor() {
		return (
			<Input
				fluid
				size='mini'
				action={<Button onClick={ this.onNameUpdateClick } >Update</Button>}
				value={ this.props.user.name }
				placeholder='Enter name...'
				onChange={ this.onNameChange }
			/>
		)
	}

	public onNameUpdateClick() {
		this.props.userUpdate().then(() => {
			this.props.router.push('/')
		})
	}

	public onNameChange(e: React.FormEvent<HTMLInputElement>) {
		this.props.userNameChange(e.currentTarget.value)
	}

	public onEditNameClick() {
		this.props.userNameEditorEnable()
	}

	public onPrevPageClick() {
		this.props.userSearch('prev')
	}

	public onNextPageClick() {
		this.props.userSearch('next')
	}

	public renderUser() {
		return (
			<Card centered>
				<ProgressiveImage src={ this.props.user.avatarUrl } placeholder={ placeholder }>
					{ (src: string) => <Image width='100%' src={ src } /> }
				</ProgressiveImage>
				<Card.Content>
					<Card.Header>
						{ this.props.isUserNameEditorEnabled ? this.renderNameEditor() : this.renderName() }
					</Card.Header>
					<Card.Description>ID: { this.props.user.id }</Card.Description>
				</Card.Content>
			</Card>
		)
	}

	public renderLoader() {
		return <Loader>Loading</Loader>
	}

}

const mapStateToProps = (state: IState, { params }: WithRouterProps) => {
	return {
		...state.userUpdatePage,
		id: params.id
	}
}

const mapDispatchToProps = { userNameEditorEnable, userReview, userUpdate, userNameChange }

export default connect(mapStateToProps, mapDispatchToProps)(UserUpdatePage)
