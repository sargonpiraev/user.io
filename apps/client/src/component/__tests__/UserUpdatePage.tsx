import 'core-js/es6/map'
import 'core-js/es6/set'
import 'raf/polyfill'

import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { UserUpdatePage } from './../UserUpdatePage'

test('userReview prop should be called after render', () => {
	const component = renderer.create(<UserUpdatePage id='1' userReview={ jest.fn() } />)
	expect(component.root.instance.props.userReview.mock.calls.length).toBe(1)
	expect(component.root.instance.props.userReview.mock.calls[0][0]).toBe(1)
})

test('user is not defined', () => {
	const tree = renderer.create(<UserUpdatePage id='1' userReview={ jest.fn() } />).toJSON()
	expect(tree).toMatchSnapshot()
})

test('user is defined', () => {
	const user = { id: 1, name: 'John Doe', avatarUrl: 'https://www.google.com/favicon.ico' }
	const tree = renderer.create(
		<UserUpdatePage
			id='1'
			user={ user }
			userReview={ jest.fn() }
		/>
	).toJSON()
	expect(tree).toMatchSnapshot()
})
