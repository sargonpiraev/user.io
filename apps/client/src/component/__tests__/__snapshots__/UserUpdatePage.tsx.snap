// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user is defined 1`] = `
<div
  className="ui container"
  style={
    Object {
      "padding": "1rem 0",
    }
  }
>
  <h1
    className="ui header"
  >
    User update page
  </h1>
  <div
    className="ui divider"
  />
  <div
    className="ui centered card"
    href={undefined}
    onClick={[Function]}
  >
    <img
      className="ui image"
      src="https://www.google.com/favicon.ico"
      width="100%"
    />
    <div
      className="content"
    >
      <div
        className="header"
      >
        <div>
          <span
            id="name"
          >
            John Doe
          </span>
          Â 
          <i
            aria-hidden="true"
            className="pencil icon"
            onClick={[Function]}
          />
        </div>
      </div>
      <div
        className="description"
      >
        ID: 
        1
      </div>
    </div>
  </div>
</div>
`;

exports[`user is not defined 1`] = `
<div
  className="ui container"
  style={
    Object {
      "padding": "1rem 0",
    }
  }
>
  <h1
    className="ui header"
  >
    User update page
  </h1>
  <div
    className="ui divider"
  />
  <div
    className="ui text loader"
  >
    Loading
  </div>
</div>
`;
